File: demos/jans-chip/app/src/main/java/io/jans/chip/keyGen/DPoPProofFactory.java
package io.jans.chip.keyGen;

import java.security.InvalidAlgorithmParameterException;
import java.security.NoSuchAlgorithmException;
import java.security.NoSuchProviderException;
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;

public class DPoPProofFactory {

    private static DPoPProofFactory single_instance = null;
    private DPoPProofFactory(){}

    public static synchronized DPoPProofFactory getInstance() {
        if (single_instance == null)
            single_instance = new DPoPProofFactory();

        return single_instance;
    }
    /**
     * The function `issueDPoPJWTToken` generates a DPoP (Distributed Proof of Possession) JWT (JSON Web Token) with the
     * specified HTTP method and request URL.
     *
     * @param httpMethod The `httpMethod` parameter represents the HTTP method used in the request, such as "GET", "POST",
     * "PUT", etc.
     * @param requestUrl The `requestUrl` parameter is the URL of the HTTP request that you want to issue a DPoP JWT token
     * for. It represents the target resource or endpoint that you want to access.
     * @return The method is returning a JWT (JSON Web Token) token.
     */
    public String issueDPoPJWTToken(String httpMethod, String requestUrl) throws InvalidAlgorithmParameterException, NoSuchAlgorithmException, NoSuchProviderException {

        Map<String, Object> headers = new HashMap<>();
        headers.put("typ", "dpop+jwt");
        headers.put("alg", "RS256");
        headers.put("jwk", KeyManager.getPublicKeyJWK(KeyManager.getInstance().getPublicKey()).getRequiredParams());

        Map<String, Object> claims = new HashMap<>();
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            LocalDateTime now = LocalDateTime.now();
            Instant instant = now.atZone(ZoneId.systemDefault()).toInstant();
            Date iat = Date.from(instant);
            claims.put("iat", iat);
        }

        claims.put("jti", UUID.randomUUID().toString());
        claims.put("htm", httpMethod); //POST
        claims.put("htu", requestUrl); //issuer

        String token = Jwts.builder()
                .setHeaderParams(headers)
                .setClaims(claims)
                .signWith(SignatureAlgorithm.RS256, KeyManager.getInstance().getPrivateKey())
                .compact();
        return token;
    }

    /**
     * The function generates a JWT token with specified claims and signs it using an RSA256 algorithm.
     *
     * @param claims A map containing the claims to be included in the JWT token. Claims are key-value pairs that provide
     * information about the token, such as the issuer, subject, expiration time, etc.
     * @return The method returns a JWT (JSON Web Token) token as a String.
     */
    public String issueJWTToken(Map<String, Object> claims) throws InvalidAlgorithmParameterException, NoSuchAlgorithmException, NoSuchProviderException {

        Map<String, Object> headers = new HashMap<>();
        headers.put("typ", "dpop+jwt");
        headers.put("alg", "RS256");
        headers.put("jwk", KeyManager.getPublicKeyJWK(KeyManager.getInstance().getPublicKey()).getRequiredParams());

        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            LocalDateTime now = LocalDateTime.now();
            Instant instant = now.atZone(ZoneId.systemDefault()).toInstant();
            Date iat = Date.from(instant);
            claims.put("iat", iat);
        }

        claims.put("jti", UUID.randomUUID().toString());

        String token = Jwts.builder()
                .setHeaderParams(headers)
                .setClaims(claims)
                .signWith(SignatureAlgorithm.RS256, KeyManager.getInstance().getPrivateKey())
                .compact();
        return token;
    }
}

File: demos/jans-chip/android/app/src/main/java/io/jans/chip/factories/DPoPProofFactory.kt
package io.jans.chip.factories

import android.os.Build
import com.nimbusds.jose.JWSObject
import com.nimbusds.jwt.JWTClaimsSet
import io.jans.chip.factories.KeyManager.Companion.getPublicKeyJWK
import io.jans.chip.utils.AppConfig
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import java.security.InvalidAlgorithmParameterException
import java.security.NoSuchAlgorithmException
import java.security.NoSuchProviderException
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.Date
import java.util.UUID

class DPoPProofFactory {
    companion object {
        var keyManager: KeyManager = KeyManager()

        @Throws(
            InvalidAlgorithmParameterException::class,
            NoSuchAlgorithmException::class,
            NoSuchProviderException::class
        )
        fun issueDPoPJWTToken(httpMethod: String, requestUrl: String): String? {

            val headers: MutableMap<String, Any> =
                mutableMapOf()
            headers["typ"] = "dpop+jwt"
            headers["alg"] = "RS256"
            val requiredParams = getPublicKeyJWK(KeyManager.getPublicKey())?.requiredParams
            headers["jwk"] = requiredParams!!

            val claims: MutableMap<String, Any> = HashMap()
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                val now = LocalDateTime.now()
                val instant =
                    now.atZone(ZoneId.systemDefault()).toInstant()
                val iat = Date.from(instant)
                claims["iat"] = iat
            }
            claims["jti"] = UUID.randomUUID().toString()
            claims["htm"] = httpMethod //POST
            claims["htu"] = requestUrl //issuer
            return Jwts.builder()
                .setHeaderParams(headers)
                .setClaims(claims)
                .signWith(SignatureAlgorithm.RS256, KeyManager.getPrivateKey())
                .compact()
        }

        @Throws(
            InvalidAlgorithmParameterException::class,
            NoSuchAlgorithmException::class,
            NoSuchProviderException::class
        )
        fun issueJWTToken(claims: MutableMap<String?, Any?>): String? {
            val headers: MutableMap<String, Any> =
                java.util.HashMap()
            headers["typ"] = "jwt"
            headers["alg"] = "RS256"
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                val now = LocalDateTime.now()
                val instant =
                    now.atZone(ZoneId.systemDefault()).toInstant()
                val iat = Date.from(instant)
                claims["iat"] = iat
            }
            claims["jti"] = UUID.randomUUID().toString()
            return Jwts.builder()
                .setHeaderParams(headers)
                .setClaims(claims)
                .signWith(SignatureAlgorithm.RS256, KeyManager.getPrivateKey())
                .compact()
        }

        fun getClaimsFromSSA(): JWTClaimsSet {
            val jwsObject: JWSObject = JWSObject.parse(AppConfig.SSA)
            val claims: JWTClaimsSet = JWTClaimsSet.parse(jwsObject.getPayload().toJSONObject());
            return claims
        }
    }
}